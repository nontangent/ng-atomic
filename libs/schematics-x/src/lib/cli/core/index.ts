export * from './key-event-handler';
