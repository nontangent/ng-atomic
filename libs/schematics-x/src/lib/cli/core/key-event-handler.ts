type Fn = () => void;

export class KeyEventHandler {
  private keyNameMap = new Map<string, Fn>();
  private defaultFn: Fn;

  registerDefault(fn: Fn) {
    this.defaultFn = fn;
  }

  registerByName(name: string, fn: Fn) {
    this.keyNameMap.set(name, fn);
  }

  handle(event: {name?: string}) {
    if (event?.name) {
      return this.keyNameMap.get(event.name)?.();
    } else {
      return this.defaultFn?.();
    }
  }
}

export const KEY_EVENT_HANDLER = '__keyEventHandler__';

export function KeyBinding(param?: {name?: string}) {
  return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
    const handler = target[KEY_EVENT_HANDLER] ??= new KeyEventHandler();
    const originalMethod = descriptor.value;
    const fn = function (...args: any[]) {
      return originalMethod.apply(this, args);
    };

    if (param?.name) {
      handler.registerByName(param.name, fn);
    } else {
      handler.registerDefault(fn);
    }
    
    return descriptor;
  };
};
