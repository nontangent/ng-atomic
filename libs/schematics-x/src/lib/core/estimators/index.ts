export * from './output-file-entry';
export * from './output-file-paths';
export * from './related-file-paths';
