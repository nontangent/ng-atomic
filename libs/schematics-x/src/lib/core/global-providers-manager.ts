import { Provider } from "@nx-ddd/core/di/interface/provider";

export class GlobalProvidersManager {
  private static _providers: Provider[] = [];

  static get providers(): Provider[] {
    return this._providers;
  }

  static register(providers: Provider[]) {
    this._providers = [...this._providers, ...providers];
  }
}
