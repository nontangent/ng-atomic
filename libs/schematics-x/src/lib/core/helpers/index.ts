export * from './distance';
export * from './helpers';
