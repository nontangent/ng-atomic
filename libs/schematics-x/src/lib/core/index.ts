export * from './decorators';
export * from './global-providers-manager';
