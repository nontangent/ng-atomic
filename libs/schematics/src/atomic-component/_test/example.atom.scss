@use 'scoped-var' as * with ($host: 'example');
@use 'atomic/atom' as *;

:host {
  @include atom($host);
}
