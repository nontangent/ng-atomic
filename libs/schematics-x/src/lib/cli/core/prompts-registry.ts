import { Inject, Injectable, InjectionToken, Injector, NxModuleFactory, Optional, resolveAndCreate } from "@nx-ddd/core";
import { Provider } from "@nx-ddd/core/di/interface/provider";
import { registerPrompt } from "inquirer";
import { GlobalProvidersManager } from "../../core";
import { InquirerAdapter } from "../adapters/inquirer";
import { Prompter, PrompterModule } from "../prompters";
import { SX_SCREEN } from "../renderer";

export class PrompterFactory {
  constructor(
    private prompterType: any,
    private nxModuleDef: any,
  ) { }
  
  create(injector: Injector) {
    const factory = new NxModuleFactory(this.nxModuleDef);
    const nxModuleRef = factory.create(injector);
    return nxModuleRef.injector.get(this.prompterType);
  }
}

export function buildPrompterFactory(prompterType, nxModuleDef) {
  return new PrompterFactory(prompterType, nxModuleDef);
}

export type PromptRegister = () => {name: string, prompterType: any, nxModuleDef: any};
export const PROMPTS_REGISTER = new InjectionToken<PromptRegister>('PromptsRegister');

export function createInjector(providers: Provider[] = [], parentInjector?: Injector) {
  return resolveAndCreate(providers, parentInjector);
}


@Injectable()
export class PromptsRegistry {
  constructor(
    protected injector: Injector,
    @Optional() @Inject(PROMPTS_REGISTER) registers?: PromptRegister[]
  ) {
    registers ??= []
    registers.forEach((register) => {
      const { name, prompterType, nxModuleDef } = register();
      this.register(name, prompterType, nxModuleDef);
    });
  }

  register(name: string, PrompterType: any, PrompterModule) {
    const factory = buildPrompterFactory(PrompterType, PrompterModule);

    registerPrompt(name, InquirerAdapter((screen) => {
      const injector = createInjector([{ provide: SX_SCREEN, useValue: screen }], this.injector);
      return factory.create(injector);
    }));
  }
}

// GlobalProvidersManager.register([
//   {
//     provide: PROMPTS_REGISTER,
//     useValue: () => ({name: 'suggest', prompterType: Prompter, nxModuleDef: PrompterModule}),
//     multi: true,
//   },
//   {
//     provide: PROMPTS_REGISTER,
//     useValue: () => ({name: 'suggest', prompterType: HistoryInputPrompter, nxModuleDef: HistoryInputModule}),
//     multi: true,
//   },
// ]);