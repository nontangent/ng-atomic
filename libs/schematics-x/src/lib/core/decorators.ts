import { Injectable as _Injectable, Injector } from '@nx-ddd/core';

export function Injectable(params?: { providedIn: 'root'}) {
  return (target: any) => {
    if (params?.providedIn === 'root') {
      target.providedIn = 'root';
      const deps = Reflect.getMetadata("design:paramtypes", target) || [];
      target['Θfac'] = () => target['Θins'] ??= new target(...deps.map(dep => dep['Θfac']()));
    }
    return _Injectable()(target);
  }
}

export class RootScopedInjector extends Injector {
  get<T>(token: any) {
    if(token.providedIn === 'root') {
      return token['Θfac']();
    }
  }
}
