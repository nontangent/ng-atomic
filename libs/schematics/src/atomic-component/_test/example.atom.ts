import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'atoms-example',
  templateUrl: './example.atom.html',
  styleUrls: ['./example.atom.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ExampleAtom implements OnInit {

  constructor() { }

  ngOnInit(): void {
  }

}
