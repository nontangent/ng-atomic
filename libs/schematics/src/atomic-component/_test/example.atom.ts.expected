import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'atoms-example',
  templateUrl: './example.atom.html',
  styleUrls: ['./example.atom.scss']
})
export class ExampleAtom implements OnInit {

  constructor() { }

  ngOnInit(): void {
  }

}
